### DISCUSSION TITLE
How to integrate different buildpack versions into a single builder ?

### DISCUSSION THREAD

**Discussion by AshwinSridharan0410** [paketo-buildpacks/feedback#255] [Node.js Team]
I am using a builder called abc, I want node versions 14,16,18,19,20 to participate in my builder.
I want to integrate https://github.com/paketo-buildpacks/nodejs/releases/tag/v2.2.0 and https://github.com/paketo-buildpacks/nodejs/releases/tag/v1.8.0 in my builder, is it possible ?

https://github.com/paketo-buildpacks/nodejs/releases/tag/v2.2.0 - this contains node version14,16,18
https://github.com/paketo-buildpacks/nodejs/releases/tag/v1.8.0 - this contains node 16,18,19,20

contents of my builder.toml

[[buildpacks]]
uri = "docker://xyz/paketo-buildpacks/nodejs:1.8.0"
version = "1.8.0"
[[buildpacks]]
uri = "docker://xyz/paketo-buildpacks/nodejs:2.2.0"
version = "2.2.0"

[[order]]

[[order.group]]
id = "paketo-buildpacks/nodejs"
version = "1.8.0"

[[order]]

[[order.group]]
id = "paketo-buildpacks/nodejs"
version = 2.2.0"

[stack]
build-image = "xyz/stacks/build-stack/base:0.0.2"
id = "io.buildpacks.stacks.jammy"
run-image = "xyz/stacks/run-stack/base:0.0.2"
run-image-mirrors = []

This wont work , to my knowledge as the buildpack is having the same id but different versions.

How to ensure that my usecase can be worked out ?
Thank you

**Comment by AshwinSridharan0410** [comment]
When I tried to build my app with this builder image, it was showing this 
failed to satisfy "node" dependency version constraint "20": no compatible versions on "io.buildpacks.stacks.jammy" stack. Supported versions are: [14.21.2, 14.21.3, 16.20.0, 16.20.1, 18.16.1, 18.17.0, 19.8.1, 19.9.0]

**Comment by dmikusa** [comment]
I don't think you could do that because while you can include multiple copies of a buildpack in the builder, you'll probably end up with detection issues. I think both of those versions of Node.js are going to have the same detection criteria, so the one you list first in the order group is always going to win detection. Unless the buildpack in question is taking into consideration at detection time the version you request and if it supports that version (I don't know as I'm not super familiar with Node.js buildpacks). Maybe one of the @paketo-buildpacks/nodejs-maintainers could comment on that.

> I am using a builder called abc, I want node versions 14,16,18,19,20 to participate in my builder.

What I would suggest is to not use two versions of the buildpack. Instead to repackage the buildpack.

So you can do something like this:

1. Clone the repo for node-engine.
2. Checkout the version tag that you want of the buildpack.
3. Modify buildpack.toml to include all of the versions you want to support.
4. Package node-engine from your modified buildpack.toml.
5. Include that buildpack in your builder.

This sides steps the whole detection issue cause you just have one buildpack. You may not even need a builder at that point, unless you have other reasons for it, you could probably just do something like `pack build -b custom-node-engine -b paketo-buildpacks/nodejs ...` and your custom buildpack runs first so it'll execute and provide Node.js.

**Reply by AshwinSridharan0410** [reply]
Thanks @dmikusa , let me try that and check
